GPU/Common/VertexDecoderX86.cpp

```diff
@@ -210,63 +230,49 @@ JittedVertexDecoder VertexDecoderJitCache::Compile(const VertexDecoder &dec, int
+	if (prescaleStep) {
+		// uvScalePtrReg should point to gstate_c.uv, or wherever the UV scale we want to use is located.
+		MOVUPS(fpScaleOffsetReg, MatR(uvScalePtrReg));
+		if ((dec.VertexType() & GE_VTYPE_TC_MASK) == GE_VTYPE_TC_8BIT) {
+			MOV(PTRBITS, R(tempReg2), ImmPtr(&by128_11));
+			MULPS(fpScaleOffsetReg, MatR(tempReg2));
+		} else if ((dec.VertexType() & GE_VTYPE_TC_MASK) == GE_VTYPE_TC_16BIT) {
+			MOV(PTRBITS, R(tempReg2), ImmPtr(&by32768_11));
+			MULPS(fpScaleOffsetReg, MatR(tempReg2));
 		}
 	}
 
 	// Add code to convert matrices to 4x4.
 	// Later we might want to do this when the matrices are loaded instead.
+	// Can't touch fpScaleOffsetReg (XMM0) in here!
 	if (dec.skinInDecode) {
 		MOV(PTRBITS, R(tempReg1), ImmPtr(&threeMasks));
-		MOVAPS(XMM4, MatR(tempReg1));
+		MOVAPS(XMM5, MatR(tempReg1));
 		MOV(PTRBITS, R(tempReg1), ImmPtr(&aOne));
-		MOVUPS(XMM5, MatR(tempReg1));
+		MOVUPS(XMM6, MatR(tempReg1));
 		MOV(PTRBITS, R(tempReg1), ImmPtr(gstate.boneMatrix));
 		MOV(PTRBITS, R(tempReg2), ImmPtr(bones));
 		for (int i = 0; i < dec.nweights; i++) {
-			MOVUPS(XMM0, MDisp(tempReg1, (12 * i) * 4));
-			MOVUPS(XMM1, MDisp(tempReg1, (12 * i + 3) * 4));
-			MOVUPS(XMM2, MDisp(tempReg1, (12 * i + 3 * 2) * 4));
-			MOVUPS(XMM3, MDisp(tempReg1, (12 * i + 3 * 3) * 4));
-			ANDPS(XMM0, R(XMM4));
-			ANDPS(XMM1, R(XMM4));
-			ANDPS(XMM2, R(XMM4));
-			ANDPS(XMM3, R(XMM4));
-			ORPS(XMM3, R(XMM5));
-			MOVAPS(MDisp(tempReg2, (16 * i) * 4), XMM0);
-			MOVAPS(MDisp(tempReg2, (16 * i + 4) * 4), XMM1);
-			MOVAPS(MDisp(tempReg2, (16 * i + 8) * 4), XMM2);
-			MOVAPS(MDisp(tempReg2, (16 * i + 12) * 4), XMM3);
-		}
-	}
-
-	// Keep the scale/offset in a few fp registers if we need it.
-	// TODO: Read it from an argument pointer instead of gstate_c.uv.
-	if (prescaleStep) {
-		MOV(PTRBITS, R(tempReg1), ImmPtr(&gstate_c.uv));
-		MOVUPS(fpScaleOffsetReg, MatR(tempReg1));
-		if ((dec.VertexType() & GE_VTYPE_TC_MASK) == GE_VTYPE_TC_8BIT) {
-			MOV(PTRBITS, R(tempReg2), ImmPtr(&by128_11));
-			MULPS(fpScaleOffsetReg, MatR(tempReg2));
-		} else if ((dec.VertexType() & GE_VTYPE_TC_MASK) == GE_VTYPE_TC_16BIT) {
-			MOV(PTRBITS, R(tempReg2), ImmPtr(&by32768_11));
-			MULPS(fpScaleOffsetReg, MatR(tempReg2));
+			MOVUPS(XMM1, MDisp(tempReg1, (12 * i) * 4));
+			MOVUPS(XMM2, MDisp(tempReg1, (12 * i + 3) * 4));
+			MOVUPS(XMM3, MDisp(tempReg1, (12 * i + 3 * 2) * 4));
+			MOVUPS(XMM4, MDisp(tempReg1, (12 * i + 3 * 3) * 4));
+			ANDPS(XMM1, R(XMM5));
+			ANDPS(XMM2, R(XMM5));
+			ANDPS(XMM3, R(XMM5));
+			ANDPS(XMM4, R(XMM5));
+			ORPS(XMM4, R(XMM6));
+			MOVAPS(MDisp(tempReg2, (16 * i) * 4), XMM1);
+			MOVAPS(MDisp(tempReg2, (16 * i + 4) * 4), XMM2);
+			MOVAPS(MDisp(tempReg2, (16 * i + 8) * 4), XMM3);
+			MOVAPS(MDisp(tempReg2, (16 * i + 12) * 4), XMM4);
 		}
 	}
 
 	// Let's not bother with a proper stack frame. We just grab the arguments and go.
-	JumpTarget loopStart = GetCodePtr();
+	JumpTarget loopStart = NopAlignCode16();
```

Comment:
It'd be a bit of a pain, but I wonder if it'd be better to measure the start size, and then if this wasn't aligned, set a private member to that size and recurse.  Then we could align the start of the function without any padding.  Not sure if worth it, though.

-[Unknown]

Reply:
Since the multi-byte nops we use are crunched very fast even by old CPUs, and it's outside the loop, I don't think it'll even be possible to measure the difference, probably not even if we move to 32-byte alignment.